packages/components/src/utils/mixins/FocusTrapMixin.ts

```diff
@@ -0,0 +1,318 @@
+      }
+    }
+
+    /**
+     * Calculates the next index for the focus trap.
+     */
+    private calculateNextIndex(currentIndex: number, step: number) {
+      const { length } = this.focusableElements;
+      const wrapFocus = this.shouldWrapFocus();
+
+      if (currentIndex === -1) {
+        return step > 0 ? 0 : length - 1;
+      }
+
+      let nextIndex = currentIndex + step;
+
+      if (wrapFocus) {
+        if (nextIndex < 0) nextIndex = length - 1;
+        if (nextIndex >= length) nextIndex = 0;
+      } else {
+        if (nextIndex < 0) nextIndex = 0;
+        if (nextIndex >= length) nextIndex = length - 1;
+      }
+
+      return nextIndex;
+    }
+
+    /**
+     * Finds the deepest active element within the shadow DOM.
+     */
+    private getDeepActiveElement() {
+      let host: Element | null = document.activeElement || document.body;
+      while (host instanceof HTMLElement && host.shadowRoot?.activeElement) {
+        host = host.shadowRoot.activeElement;
+      }
+
+      return host as HTMLElement || document.body;
+    }
+
+    /**
+     * Finds the index of the active element within the focusable elements.
+     */
+    private findElement(activeElement: HTMLElement) {
+      return this.focusableElements.findIndex((element) => this.isEqualFocusNode(activeElement, element));
+    }
+
+    /**
+     * Checks if the active element is the same as the given element.
+     */
+    private isEqualFocusNode(activeElement: HTMLElement, element: HTMLElement) {
+      if (activeElement.nodeType >= 0) {
+        return element.isEqualNode(activeElement) && element === activeElement;
+      }
+      return false;
+    }
+
+    /**
+     * Traps focus within the container.
+     */
+    private trapFocus(direction: boolean) {
+      if (this.focusableElements.length === 0) return;
+
+      const activeElement = this.getDeepActiveElement!() as HTMLElement;
+      const activeIndex = this.findElement(activeElement);
+
+      if (direction) {
+        this.focusTrapIndex = this.calculateNextIndex(activeIndex, -1);
+      } else {
+        this.focusTrapIndex = this.calculateNextIndex(activeIndex, 1);
+      }
+
+      const nextElement = this.focusableElements[this.focusTrapIndex];
+      if (nextElement) {
+        nextElement.focus();
+      }
+    }
+
+    /**
+     * Handles the keydown event to trap focus within the container.
+     */
+    private handleKeydown(event: KeyboardEvent) {
+      if (!this.enableFocusTrap || !this.focusableElements.length) {
+        return;
+      }
+
+      if (event.key === 'Tab') {
+        event.preventDefault();
+        this.trapFocus(event.shiftKey);
+      }
+    }
+
+    override connectedCallback() {
+      super.connectedCallback();
+      this.addEventListener('keydown', this.handleKeydown);
+    }
+
+    override disconnectedCallback() {
+      super.disconnectedCallback();
+      this.removeEventListener('keydown', this.handleKeydown);
+    }
+  }
```

Comment:
```suggestion
override constructor(){
  super();
  this.addEventListener('keydown', this.handleKeydown)
  }
}
```

This will take care of registering as well as removing the listener for cleanup.

https://hawkticehurst.com/2023/11/you-are-probably-using-connectedcallback-wrong/